photon-api/src/main/scala/com/linkedin/photon/ml/transformers/GameTransformer.scala

```diff
@@ -0,0 +1,308 @@
+    setDefault(logDataAndModelStats, false)
+    setDefault(spillScoresToDisk, false)
+  }
+
+  /**
+   * Verify that the interactions between individual parameters are valid.
+   *
+   * @note In Spark, interactions between parameters are checked by
+   *       [[org.apache.spark.ml.PipelineStage.transformSchema()]]. Since we do not use the Spark pipeline API in
+   *       Photon-ML, we need to have this function to check the interactions between parameters.
+   *
+   * @throws IllegalArgumentException if a required parameter is missing or a validation check fails
+   */
+  protected[transformers] def validateParams(): Unit = {
+
+    // Just need to check that these parameters are explicitly set
+    getRequiredParam(model)
+  }
+
+  /**
+   * Return the user-supplied value for a required parameter. Used for mandatory parameters without default values.
+   *
+   * @tparam T The type of the parameter
+   * @param param The parameter
+   * @return The value associated with the parameter
+   * @throws IllegalArgumentException if no value is associated with the given parameter
+   */
+  private def getRequiredParam[T](param: Param[T]): T =
+    get(param)
+      .getOrElse(throw new IllegalArgumentException(s"Missing required parameter ${param.name}"))
+
+  //
+  // GameTransformer functions
+  //
+
+  /**
+   * Transform a [[DataFrame]] of data samples to include scores computed using the sample feature vectors and the
+   * currently set model.
+   *
+   * @param data Input [[DataFrame]] of samples
+   * @return Scored data samples
+   */
+  def transform(data: DataFrame): ModelDataScores = {
+
+    validateParams()
+
+    val randomEffectTypes = getRequiredParam(model)
+      .toMap
+      .values
+      .flatMap {
+        case rem: RandomEffectModel => Seq(rem.randomEffectType)
+        case _ => Seq()
+      }
+      .toSet
+    val featureShards = getRequiredParam(model)
+      .toMap
+      .values
+      .map {
+        case fem: FixedEffectModel => fem.featureShardId
+        case rem: RandomEffectModel => rem.featureShardId
+      }
+      .toSet
+
+    val gameDataSet = Timed("Preparing GAME data set") {
+      prepareGameDataSet(data, randomEffectTypes, featureShards)
+    }
+
+    if (getOrDefault(logDataAndModelStats)) {
+      logGameDataSet(gameDataSet, randomEffectTypes)
+      logger.debug(s"GAME model summary:\n${getRequiredParam(model).toSummaryString}")
+    }
+
+    val scores = Timed("Computing scores") {
+      scoreGameDataSet(gameDataSet)
+    }
+
+    gameDataSet.unpersist()
+
+    Timed("Evaluating scores") {
+      get(validationEvaluators).foreach(
+        _.foreach { evaluatorType =>
+          val evaluationMetricValue = evaluateScores(evaluatorType, gameDataSet, scores)
+          logger.info(s"Evaluation metric value on scores with $evaluatorType: $evaluationMetricValue")
+        })
+    }
+
+    // TODO: Instead, we should merge the scores back into the DataFrame in a new column (at least optionally)
```

Comment:
Yeah. This would also be more in the spirit of the `Transformer` pattern in spark.ml.

Reply:
Yes, I gave it a quick go but it became a bit of an ordeal so I decided to leave it for a later commit.